include/dsn/tool-api/rpc_address.h

```diff
@@ -0,0 +1,243 @@
+                      "to keep it sizeof(uint64_t)");
+    }
+
+    rpc_address(const char *host, uint16_t port) { assign_ipv4(host, port); }
+
+    void assign_ipv4(uint32_t ip, uint16_t port)
+    {
+        clear();
+        _addr.v4.type = HOST_TYPE_IPV4;
+        _addr.v4.ip = ip;
+        _addr.v4.port = port;
+    }
+
+    void assign_ipv4(const char *host, uint16_t port)
+    {
+        clear();
+        _addr.v4.type = HOST_TYPE_IPV4;
+        _addr.v4.ip = rpc_address::ipv4_from_host(host);
+        _addr.v4.port = port;
+    }
+
+    void assign_ipv4_local_address(const char *network_interface, uint16_t port)
+    {
+        clear();
+        _addr.v4.type = HOST_TYPE_IPV4;
+        _addr.v4.ip = rpc_address::ipv4_from_network_interface(network_interface);
+        _addr.v4.port = port;
+    }
+
+    void assign_uri(const char *host_uri);
+    void assign_group(const char *name);
+
+    rpc_address clone() const;
+    const char *to_string() const;
+
+    std::string to_std_string() const { return std::string(to_string()); }
+    bool from_string_ipv4(const char *s)
+    {
+        clear();
+        std::string str = std::string(s);
+        auto pos = str.find_last_of(':');
+        if (pos == std::string::npos)
+            return false;
+        else {
+            auto host = str.substr(0, pos);
+            auto port_str = str.substr(pos + 1);
+            char *p = nullptr;
+            long port = ::strtol(port_str.data(), &p, 10);
+            if (*p != 0) // bad string
+                return false;
+            if (port <= 0 || port > UINT16_MAX) // out of range
+                return false;
+            assign_ipv4(host.c_str(), (uint16_t)port);
+            return true;
+        }
+    }
+
+    uint64_t &value() { return _addr.value; }
+
+    dsn_host_type_t type() const { return (dsn_host_type_t)_addr.v4.type; }
+    uint32_t ip() const { return (uint32_t)_addr.v4.ip; }
+    uint16_t port() const { return (uint16_t)_addr.v4.port; }
+    void set_port(uint16_t port) { _addr.v4.port = port; }
+    rpc_group_address *group_address() const
+    {
+        return (rpc_group_address *)(uintptr_t)_addr.group.group;
+    }
+    rpc_uri_address *uri_address() const { return (rpc_uri_address *)(uintptr_t)_addr.uri.uri; }
+    bool is_invalid() const { return _addr.v4.type == HOST_TYPE_INVALID; }
+    void set_invalid() { clear(); }
+
+    bool operator==(::dsn::rpc_address r) const
+    {
+        if (type() != r.type())
+            return false;
+
+        switch (type()) {
+        case HOST_TYPE_IPV4:
+            return ip() == r.ip() && _addr.v4.port == r.port();
+        case HOST_TYPE_URI:
+            return strcmp(to_string(), r.to_string()) == 0;
+        case HOST_TYPE_GROUP:
+            return _addr.group.group == r._addr.group.group;
+        default:
+            return true;
+        }
+    }
+
+    bool operator!=(::dsn::rpc_address r) const { return !(*this == r); }
```

Comment:
Remove this. Compiler automatically generates the inequality operator if equality has defined.